core/src/main/java/com/google/errorprone/bugpatterns/AbortInOvercatch.java

```diff
@@ -0,0 +1,138 @@
+
+import static com.sun.source.tree.Tree.Kind.EXPRESSION_STATEMENT;
+
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.TreeInfo;
+import com.sun.source.tree.CatchTree;
+import com.sun.source.tree.StatementTree;
+import com.sun.source.tree.TryTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.ExpressionStatementTree;
+
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+
+import java.util.List;
+
+/**
+ * A bug checker for abort in exception overcatch. See the following example: 
+ *
+ * <pre>
+ * try {
+ *   // stmt 1 throws exception B
+ *   // stmt 2 throws exception A
+ * } catch (Throwable t) {
+ *   // more than one unique exceptions fall into this catch -- overcatch
+ *   abort(); 
+ * }
+ * </pre>
+ *
+ * While it is a best practice to catch the precise exceptions, abort or System.exit()
+ * in exception over-catches are particularly dangerous as it might accidentally
+ * bring down the cluster on unexpected scenarios. 
+ * 
+ * Because it is difficult to check for the exceptions each statement throws in 
+ * AST, we take the liberty to simplify our implementation: we simply check if the
+ * catch block catches "Throwable" or "Exception", and then calls abort(). 
+ * 
+ * For more detail, refer to the paper:
+ * "Simple Testing Can Prevent Most Critical Failures: 
+ *  An Analysis of Production Failures in Distributed Data-intensive Systems"
+ *  Yuan et al. Proceedings of the 11th Symposium on Operating Systems Design 
+ *  and Implementation (OSDI), 2014
+ *
+ * @author yuan@eecg.utoronto.ca (Ding Yuan)
+ */
+@BugPattern(name = "AbortInOvercatch",
+    summary = "Aborting the system after catching Throwable or Exception. "
+            + "Because these exceptions are high-level exceptions, this might accidentally"
+            + " take down the systems on unexpected exceptions.",
+    explanation = "Aborting the system in Throwable or Exception is bad practice.",
+    category = JDK, maturity = MATURE, severity = ERROR)
+public class AbortInOvercatch extends BugChecker implements TryTreeMatcher {
+  @Override
+  public Description matchTry (TryTree tree, VisitorState state) {
+    List<? extends CatchTree> catchList = tree.getCatches();
+    if (catchList == null || catchList.size() == 0) {
+      // TODO: this try block does not have a catch, we should further check the 
+      // finally block!
+      return Description.NO_MATCH;
+    }
+ 
+    CatchTree lastCatch = catchList.get(tree.getCatches().size() - 1);
+    if (overcatch(lastCatch, state)) {
+      if (abortInCatch(lastCatch, state)) {
+        LineMap lineMap = state.getPath().getCompilationUnit().getLineMap();
+
+        return describeMatch(lastCatch);
+      }
+    }
+    return Description.NO_MATCH;
+  }
+
+  private boolean overcatch(CatchTree catchTree, VisitorState state) {    
+    String caughtException = catchTree.getParameter().getType().toString();
```

Comment:
I would not do string matching here, it's too brittle.  You can use the symbol table and the types helper to do real type matching:
```
Types types = state.getTypes();
Symtab symtab = state.getSymtab();
Type paramType = ASTHelpers.getType(catchTree.getParameter().getType());
return (types.isSameType(paramType, symtab.throwableType) || types.isSameType(paramType, symtab.exceptionType));
```